keps/sig-auth/2843-request-ctx-in-token-review/README.md

```diff
@@ -0,0 +1,109 @@
+  - [Version Skew Strategy](#version-skew-strategy)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes enhancing the TokenReview API to convey additional request
+context to the backing Identity Provider (IDP).
+
+## Motivation
+
+Today, certain request attributes that are known to kube-apiserver may be
+relevant to the backend of a TokenReview webhook. For example:
+
+* Was the request made over https?
+* What are the IP addresses of the peer and origin of the request?
+
+These attributes can be used to enhance audit logs in the IDP, determine access
+levels (such as whether the request originates from a trusted network), or
+enforce security policy centrally (such as disallowing bearer token
+authentication over insecure protocols).
+
+### Goals
+
+* Forward additional request metadata to TokenReview.
+
+### Non-Goals
+
+* Forward resource metadata to TokenReview: The object of the request (e.g.
+  which verb, which namespace, which pod) should remain unknown to the IDP.
+* Forward addition request metadata to SubjectAccessReview or AdmissionReview:
+  That's what User.ExtraInfo is for.
+
+## Proposal
+
+We will extend TokenReviewSpec to add an additional field:
+
+```golang
+
+type TokenReviewSpec struct {
+  ...
+
+  // RequestAttributes is a collection of request attributes that may be relevant
+  // to a TokenReview backend. These attributes should generally be stable over
+  // multiple requests.
+  //
+  // Note: TokenReview webhook backends must handle absences of this field or
+  // subfields of the RequestAttribute object. This data may be absent when the
+  // API Server initiating the TokenReview is old or the data was unavailable.
+  RequestAttributes RequestAttributes
+}
+...
+
+type RequestAttributes struct {  
+  // The HTTP request `Host` header value.
+  Host string
+
+  // The HTTP URL scheme, such as `http` and `https`.
+  Scheme string
+
+  // The IP address of the remote peer.
+  //
+  // Note: This is only the IP address of the originator of the request, when
+  // the network path of the request does not traverse any hops that perform
+  // SNAT. For example, if the request path traverses a simple proxy, this
+  // will be the address of the proxy.
+  PeerIP string
+
+  // Additional standard headers used by frontproxies to forward request
+  // attributes to backends. Currently, the only forwarded headers are:
+  // * X-Forwarded-For
+  // * X-Forwarded-Proto
+  ProxyHeaders map[string][]string
+}
+```
+
+These attributes will be resolved in RequestInfo filter and populated in the
+[RequestInfo] struct. The cached token authenticator will be updated to include
+request attributes in the hash key. Notably, this means that volatile
+attributes (e.g. a millisecond timestamp or trace ID) are not suitable to
+include in RequestAttributes.
+
+[RequestInfo]: https://github.com/kubernetes/kubernetes/blob/fffaadc01331cca57cafa0fc066a2a3eec23acb8/staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go#L42
+
```

Comment:
I'd be interested in a discussion (alternatives considered?) of why this shouldn't just be handled as an authenticating proxy.

Reply:
Added an alternative for front proxy. Let me know if you have any other alternatives you'd like explored in the KEP.